src/scanner/mod.rs

```diff
@@ -100,41 +103,59 @@ impl Scanner {
     ///
     /// Note: `self` must contain `self.ip`.
     async fn scan_socket(&self, socket: SocketAddr) -> io::Result<SocketAddr> {
-        match self.connect(socket).await {
-            Ok(x) => {
-                debug!(
-                    "Connection was successful, shutting down stream {}",
-                    &socket
-                );
-                match x.shutdown(Shutdown::Both) {
-                    Err(e) => debug!("Shutdown stream error {}", &e),
-                    _ => {}
-                }
-                if !self.greppable {
-                    if self.accessible {
-                        println!("Open {}", socket.to_string());
-                    } else {
-                        println!("Open {}", socket.to_string().purple());
+        let mut ret = Err(io::Error::new(
+            io::ErrorKind::Other,
+            format!("Did not connect to {}", &socket),
+        ));
+
+        for nr_try in 0..self.tries {
+            debug!("Try number: {}", nr_try + 1);
+
+            ret = match self.connect(socket).await {
```

Comment:
I think it would be clearer if the `match` did not return anything and you moved this assignment where it actually happens. This way, you could also check whether you really need the assigned value (more in my comment at the end of the function).